Make switch to data button keyboard accessible/interactive

DuduzaoMT opened the pull request:
Closes https://github.com/metabase/metabase/issues/53343

### Description

Made the button keyboard accessible and interactive by integrating it into the respective component. The behavior was updated to ensure keyboard navigation interaction, as demonstrated in the video.

### How to verify

Go to Metrics -> Select one -> Navigate to the switch button using tab -> Press Enter/Space.

Verify that you can interact with the button using the keyboard (press "Tab" to navigate to the button and "Enter" to activate it).
Confirm that the behavior works as expected without losing functionality.

### Demo

https://github.com/user-attachments/assets/b4032778-3720-4e76-981a-1b203b629ff1


### Checklist

- [X] Tests have been added/updated to cover changes in this PR

DuduzaoMT committed:
Fix #53343: Made button keyboard interactive

Made the button keyboard accessible and interactive by
integrating it into the respective component.
The behavior was updated to ensure proper keyboard
navigation and interaction.

lorem--ipsum reviewed (changes requested):
Hello @DuduzaoMT and thank you for your PR!

It needs some more work for the following reasons:
 - using the `button` role is [not recommanded](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/button_role) and misleading
 - we'd rather use Mantine's [SegmentedControl](https://mantine.dev/core/segmented-control/) to replace this entirely and benefit from built-in accessibility features (note that it would probably require some work on styling)

Please note however that we do appreciate your contribution, and particularly the fact that you implemented meaningful tests.

DuduzaoMT committed:
Merge branch 'master' into reactive-button-to-keyboard

DuduzaoMT commented:
> Hello @DuduzaoMT and thank you for your PR!
> 
> It needs some more work for the following reasons:
> 
> * using the `button` role is [not recommanded](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/button_role) and misleading
> * we'd rather use Mantine's [SegmentedControl](https://mantine.dev/core/segmented-control/) to replace this entirely and benefit from built-in accessibility features (note that it would probably require some work on styling)
> 
> Please note however that we do appreciate your contribution, and particularly the fact that you implemented meaningful tests.

Hi @lorem--ipsum !
Just to clarify, the built-in accessibility behavior for segmentedControl relies on keyboard arrow navigation, not tab/space and enter as mentioned in the issue.

That said, if this arrow-based navigation still qualifies as valid for your use case, I’d be happy to move forward with the fix! Let me know.

lorem--ipsum commented:
I think that would be acceptable, yes. I'll double check with @jessicaul for good measure.

DuduzaoMT committed:
Fix #53343: replaced Flex with SegmentedControl

Switched from Flex to SegmentedControl,
added corresponding CSS, and updated tests.

DuduzaoMT commented:
Hi @lorem--ipsum!
I added _onKeyDown_ to the suggested _segmentedControl_ to enable arrow key interaction (which is built-in in this Mantine component), and also to try to preserve the interaction proposed in the issue. If you feel it doesn't quite fit, I'm happy to remove it. I'm also available to make any other changes or corrections you might need!

DuduzaoMT commented:
> Thank you, this is looking good. Could you please also make it so clicking anywhere toggles the button? The more I think of it the more I'm convinced that it's closer, functionnally, to a switch than to a segmented control.

That makes sense.  Treating it more like a switch feels more intuitive. I can adapt the current segmented control to behave that way, or if you prefer, I can look into other components that might fit this use case better. Let me know what you'd prefer!

lorem--ipsum commented:
If using a mantine's switch involves less coding I think it's better, however if needs to be close to what we have currently in terms of styling. In summary:
 - use a SegmentedControl but modify its interactivity to match the current behavior
 - use a Switch but make it look like as close as possible to what we have currently

Does that make sense to you?

Thanks!

DuduzaoMT committed:
Fix metabase#53343: replaced SegmentedControl with Switch
Switched from SegmentedControl to Switch,
added corresponding CSS, and updated tests.

DuduzaoMT commented:
Hi @lorem--ipsum,
Thanks for the feedback — really appreciate it!
Yeah, it totally makes sense to me, so I went ahead and made the switch.
I think it turned out well — but if you have any suggestions, let me know!

lorem--ipsum commented:
Hi @DuduzaoMT ! Thank you for your efforts.

I checked out your branch, the implementation looks fine but the UI is a bit janky. It's not your fault, I think Mantine's Switch component isn't well fitted to working with icons only, to be honest. Could you please roll back to using the SegmentedControl and try my initial suggestion, which was to add an interaction so clicking anywhere on the component would toggle it? Hopefully we'll get to a point where the interactions, the accessibility and the look and feel are all up to our standards.

Thanks!

DuduzaoMT commented:
> Hi @DuduzaoMT ! Thank you for your efforts.
> 
> I checked out your branch, the implementation looks fine but the UI is a bit janky. It's not your fault, I think Mantine's Switch component isn't well fitted to working with icons only, to be honest. Could you please roll back to using the SegmentedControl and try my initial suggestion, which was to add an interaction so clicking anywhere on the component would toggle it? Hopefully we'll get to a point where the interactions, the accessibility and the look and feel are all up to our standards.
> 
> Thanks!

Hi @lorem--ipsum ! I'm currently trying to find a solution that meets that request, but Mantine's SegmentedControl has a specific way of handling state changes, which makes things a bit tricky and leads to some odd behaviors. Here’s what I’ve found so far:

.Wrapping the SegmentedControl in a div or button to make the whole area clickable.
.Building a custom component that behaves like a group of buttons.
.Sticking to the default behavior, where changes only happen via icon clicks, Enter, Space, or arrow keys (as it was in the previous commit with SegmentedControl).

If you have other suggestions, let me know.

DuduzaoMT committed:
Fix #53343: Improve SegmentedControl interaction

Made the whole control clickable while keeping keyboard access intact.

DuduzaoMT commented:
> I'd be in favor of writing a custom component, although that requires more work. A couple of guidelines if you're down for it:
> 
> * the component should live in its own file
> * it should use the proper semantic HTML foundation with `input[type="checkbox"]` for accessibility and native behavior
> * it should implement proper TypeScript interfaces for props - include options for labels, default state, onChange handlers, and custom icon components
> * it should be styled using CSS modules
> * it should have transition animations (0.2-0.3s) for smooth state changes, with proper focus/hover states for accessibility
> 
> Let me know if you need more guidance

Hi @lorem--ipsum,
Once I had some feedback, I concluded that it would be better to fix the problem with the existing segmented control, given the complexity of creating a new component from scratch.
I did my best to achieve the desired behavior, although working with the current SegmentedControl to meet all the requirements wasn’t easy, since it has a very restrictive default behavior.
I'm open to any suggestions or help to improve it. Thanks!

DuduzaoMT commented:
> Hey @DuduzaoMT I apologize for the delay, I finally found some time to check out your branch and playh with it. I think it's good to go, functionally speaking. Visually there's still a blue outline that is not aligned with other controls' outlines when active: <img alt="image" width="1607" src="https://private-user-images.githubusercontent.com/1224709/446587831-a6c846f0-55c6-4e62-b37e-1923ae0f98f7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDgxODgxMjgsIm5iZiI6MTc0ODE4NzgyOCwicGF0aCI6Ii8xMjI0NzA5LzQ0NjU4NzgzMS1hNmM4NDZmMC01NWM2LTRlNjItYjM3ZS0xOTIzYWUwZjk4ZjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDUyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA1MjVUMTU0MzQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTQ0OTRhNDI3NDZkMGQ4OTMzODVkMDFlNWIyN2Q4M2EwZmVmMWFhNWUzMDYyYTIyYThiMGNjMjYxMTExYmI3NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.FScXlmJGoJcfYOtjDCn7XSlqpEN5FoPiqgWMhJR3tqM">
> 
> Could you have a look?
> 
> Thank you

Hi @lorem--ipsum,
Are you referring to the space between the blue inner area and the blue border?
Do you want it to be the same on the right and top sides (based on the image), right?

thanks for your time!

DuduzaoMT committed:
Fix #53343: Improve SegmentedControl

Enhance SegmentedControl styles and fix minor interaction issue.

DuduzaoMT commented:
> No, sorry I was unclear, I meant the outline doesn't look like other focus/active outline on other components.

Hey @lorem--ipsum,
I think I’ve made the changes you asked for. Let me know what you think.
Thanks.

lorem--ipsum committed:
Merge branch 'master' into reactive-button-to-keyboard

DuduzaoMT committed:
Adicionado Diagrama sem nome.drawio
